Chap_API_Struct.tex

```diff
@@ -2151,9 +2151,118 @@ \subsubsection{Create the\refstruct{pmix_info_t} array of query qualifiers}
 
 
 %%%%%%%%%%%
-\section{Packing/Unpacking Types \& Structures}
+\subsection{Attribute registration structure}
+\declarestruct{pmix_regattr_t}
+
+The \refstruct{pmix_regattr_t} structure is used to register attribute support for a \ac{PMIx} function.
+
+\versionMarker{4.0}
+\cspecificstart
+\begin{codepar}
+typedef struct pmix_regattr \{
+    char *key;
```

Comment:
It's still a little unclear to me if the `key` is the macro constant or the string constant behind the macro. Meaning, for example, if we have this attribute:
```
PMIX_SERVER_TMPDIR "pmix.srvr.tmpdir"
```
Should `key = strdup("PMIX_SERVER_TMPDIR");` or should `key = strdup(PMIX_SERVER_TMPDIR)` (where the latter duplicates `pmix.srvr.tmpdir`

I can see an argument for both. I guess the question is do we expect a user to do a `strcmp(key, "PMIX_SERVER_TMPDIR")` or a `strcmp(key, PMIX_SERVER_TMPDIR)` in thier program.

Reply:
It isn't that much more work to simply accept both, frankly - i.e., we can say that the PMIx library must accept either form. The bigger question is: what should be returned?

I think the human-readable output should be the attribute name - i.e., "PMIX_SERVER"TMPDIR".

The machine-parsable output should be the attribute representation - i.e., "pmix.srvr.tmpdir" - as that is what the program would need to pass back to a PMIx API.

Make sense?

Reply:
Actually, let me correct myself - the input is always the _function_ whose attribute support is being requested, never an attribute itself. So this really just concerns the two options for output.

Reply:
On the server side, we would need an ability to translate the attribute representation string into the attribute name macro string. Not too hard to do. So we could accept either. For RM custom attributes we might only have the attribute representation to pass through the interface.

Considering the possibility of RM custom attributes being expressed through this interface, I think on the client/tool side we should try to report both the attribute name (if a translation is available) and the attribute representation. The latter will always have to exist, but the former may not for RM custom attributes.